### DISCUSSION TITLE
How to allow free variables

### DISCUSSION THREAD

**Discussion by ray-Artfax** [casadi/casadi#3225] [Q&A]
RuntimeError: Error in Function::Function for 'x_dot' [MXFunction] at .../casadi/core/function.cpp:249:
.../casadi/core/function_internal.cpp:146: Error calling MXFunction::init for 'x_dot':
.../casadi/core/mx_function.cpp:409: x_dot::init: Initialization failed since variables [gp_p, gp_a, gp_v, gp_r, trigger] are free. These symbols occur in the output expressions but you forgot to declare these as inputs. Set option 'allow_free' to allow free variables.

I am getting this error and i am not able to find documentation to allow free variables . It would be great if somebody could help. I am sorry if this is a beginner question, I am new to Casadi

**Comment by jgillis** [comment]
Are you using a library with casadi under the hood? If so, which? If not, show your Function constructor. Disallow free variables by default is a new thing in CasADi 3.6..

**Reply by ray-Artfax** [reply]
Yeah I am using the ml-Casadi(TUM) and Acados library. i suppose both their origination is from Casadi itself

**Comment by jgillis** [comment]
Can you give a full stacktrace? Normally the console should fill up with much more info than what you posted..

**Reply by ray-Artfax** [reply]
Sure, Thank You for looking into my [matter]
<img width="1152" alt="Screenshot 2023-06-12 at 11 49 35 AM" src="https://github.com/casadi/casadi/assets/78547163/4eb8845e-b67a-4d84-9a9f-0027e72ef63a">

**Reply by ray-Artfax** [reply]
How do i allow free variables in this line of code cuz this is the function thats throwing error,i guess when this code was written , it must have been written in casadi 3.5. since free variables were allowed backthen
<img width="793" alt="Screenshot 2023-06-14 at 2 18 06 PM" src="https://github.com/casadi/casadi/assets/78547163/210d3360-9786-4e21-bb7d-b99c21d9f63b">

**Comment by jgillis** [comment]
`cs.Function('x_dot',[self.x, self.u], [dyn], ['x','u'], ['x_dot'],{"allow_free":True})`

Please report this fix to the developers of the library you are using here. Free variables *may* be indicative of bugs, that's why we enforced a default strict policy.
